<?xml version="1.0"?>
<robot name="forktruck_model" xmlns:xacro="http://ros.org/wiki/xacro">
  <property name="M_PI" value="3.14159"/>
  <property name="R_wheel" value="0.23"/>
  <property name="W_wheel" value="0.08"/>
  <property name="ratio" value="0.006"/>
  <!-- Macro for forktruck body. -->
  <!-- Import Rviz colors -->
  <xacro:include filename="$(find tutorial_gazebo)/urdf/materials.xacro" />
  <property name="base_z" value="0.08" />
  
  <xacro:macro name="forktruck_body">

  <link name="base_footprint">
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <box size="0.001 0.001 0.001" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0.1" rpy="0 0 0" />
      <geometry>
       <box size="0.001 0.001 0.001" />
      </geometry>
    </collision>
  </link>

  <link name="base_link">
    <inertial>
      <mass value="0.010" />
      <origin xyz="0 0 0" />
      <inertia ixx="0.1" ixy="0.0" ixz="0.0"
        iyy="0.1" iyz="0.0" izz="0.1" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
	       <box size="0.0004 0.004 0.0004" />
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
	        <box size="0.0004 0.004 0.0004" />
      </geometry>
    </collision>
  </link>

  <joint name="base_footprint_joint" type="fixed">
    <origin xyz="-0.0 0 ${R_wheel}" rpy="0 0 ${0.0*M_PI}" />        
    <parent link="base_footprint"/>
    <child link="base_link" />
  </joint>


  <link name="base_visu_link">
    <inertial>
      <mass value="1800" />
      <origin xyz="0 0 0" />
      <inertia ixx="11" ixy="0.0" ixz="0.0"
        iyy="10" iyz="0.0" izz="15" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="${0.5*M_PI} ${0.0*M_PI} ${0.5*M_PI}" />
      <geometry>
	        <mesh filename="package://tutorial_gazebo/meshes/folktruck_base.stl" scale="${ratio} ${ratio} ${ratio}"/>
      </geometry>
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
	        <box size="0.0004 0.004 0.0004" />
      </geometry>
    </collision>
  </link>

  <joint name="base_visu_joint" type="fixed">
    <origin xyz="0 0 0.28" rpy="0 0 ${1.0*M_PI}" />        
    <parent link="base_link"/>
    <child link="base_visu_link" />
  </joint>

 <link name="f_wheel_link">
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="200" />
      <inertia ixx="5.01" ixy="0.0" ixz="0.0"
        iyy="2.1" iyz="0.0" izz="4.1" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="${0.5*M_PI} ${0.0*M_PI} ${0.5*M_PI}" />
      <geometry>
        <mesh filename="package://tutorial_gazebo/meshes/folktruck_wheel.stl" scale="${1.0*ratio} ${ratio} ${ratio}"/>
      </geometry>
         <material name="black" />
    </visual>
    <collision>
      <origin xyz="0 0 0" rpy="${0.5*M_PI} ${0.0*M_PI} ${0.5*M_PI}" />
      <geometry>
        <mesh filename="package://tutorial_gazebo/meshes/folktruck_wheel.stl" scale="${1.0*ratio} ${ratio} ${ratio}"/>
      </geometry>
    </collision>
  </link>

  <joint name="f_wheel_joint" type="fixed">
    <origin xyz="0.0 0.0 -0.040" rpy="0 0 0"/>
    <parent link="base_visu_link"/>
    <child link="f_wheel_link"/>
    <axis xyz="0 1 0"/>
  </joint>


<link name="r_wheel_link">
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="200" />
      <inertia ixx="5.01" ixy="0.0" ixz="0.0"
        iyy="2.1" iyz="0.0" izz="4.1" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="${0.5*M_PI} ${0.0*M_PI} ${0.5*M_PI}" />
      <geometry>
        <mesh filename="package://tutorial_gazebo/meshes/folktruck_wheel.stl" scale="${ratio} ${ratio} ${ratio}"/>
      </geometry>
         <material name="black" />
    </visual>

    <collision>
      <origin xyz="0 0 0" rpy="${0.5*M_PI} ${0.0*M_PI} ${0.5*M_PI}" />
      <geometry>
        <mesh filename="package://tutorial_gazebo/meshes/folktruck_wheel.stl" scale="${ratio} ${ratio} ${ratio}"/>
      </geometry>
    </collision>
  </link>

  <joint name="r_wheel_joint" type="fixed">
    <origin xyz="-2.4 0.0 -0.04" rpy="0 0 0"/>
    <parent link="base_visu_link"/>
    <child link="r_wheel_link"/>
    <axis xyz="0 1 0"/>
  </joint>


<link name="vplate">
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="10" />
      <inertia ixx="0.001" ixy="0.0" ixz="0.0"
        iyy="0.001" iyz="0.0" izz="0.001" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="${0.5*M_PI} ${0.0*M_PI} ${0.5*M_PI}" />
      <geometry>
        <mesh filename="package://tutorial_gazebo/meshes/folktruck_front.stl" scale="${ratio} ${1.5*ratio} ${ratio}"/>
      </geometry>
      <material name="grey" />
    </visual>

    <collision>
      <origin xyz="0 0 0.5" rpy="0 0 0" />
      <geometry>
        <sphere radius="0.035" />
      </geometry>
    </collision>
  </link>

  <joint name="vplate_joint" type="fixed">
    <origin xyz=".10 -0.0 0.04" rpy="0 0 0"/>
    <axis xyz="1 0 0" />
    <limit effort="2000.0" lower="0" upper="1.2" velocity="0.2"/>
    <dynamics damping="1.0"/>
    <parent link="base_visu_link"/>
    <child link="vplate"/>
  </joint>


<link name="lcynldr">
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="0.1" />
      <inertia ixx="0.001" ixy="0.0" ixz="0.0"
        iyy="0.001" iyz="0.0" izz="0.001" />
    </inertial>
    <visual>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <geometry>
        <cylinder  radius="0.05" length = "3.75"/>
      </geometry>
      <material name="orange" />
    </visual>

    <collision>
      <origin xyz="0 0 0.5" rpy="0 0 0" />
      <geometry>
        <cylinder  radius="0.0035" length = "0.2"/>
      </geometry>
    </collision>
  </link>

  <joint name="lcynldr_joint" type="fixed">
    <origin xyz="0.4 0.53 1.55" rpy="0 0 0"/>
    <parent link="vplate"/>
    <child link="lcynldr"/>
  </joint>

<link name="rcynldr">
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="0.1" />
      <inertia ixx="0.001" ixy="0.0" ixz="0.0"
        iyy="0.001" iyz="0.0" izz="0.001" />
    </inertial>
    <visual>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <geometry>
        <cylinder  radius="0.05" length = "3.75"/>
      </geometry>
      <material name="orange" />
    </visual>

    <collision>
      <origin xyz="0 0 0.5" rpy="0 0 0" />
      <geometry>
        <cylinder  radius="0.0035" length = "0.2"/>
      </geometry>
    </collision>
  </link>

  <joint name="rcynldr_joint" type="fixed">
    <origin xyz="0.4 -0.53 1.55" rpy="0 0 0"/>
    <parent link="vplate"/>
    <child link="rcynldr"/>
  </joint>


<link name="fork">
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="18" />
      <inertia ixx="0.021" ixy="0.0" ixz="0.0"
        iyy="0.021" iyz="0.0" izz="0.021" />
    </inertial>
    <visual>
      <origin xyz="0 0 0.0"  rpy="${0.5*M_PI} ${0.0*M_PI} ${0.5*M_PI}" />
      <geometry>
        <mesh filename="package://tutorial_gazebo/meshes/folktruck_fork.stl" scale="${ratio} ${ratio} ${ratio}"/>
      </geometry>
      <material name="orange" />
    </visual>

    <collision>
      <origin xyz="0 0 0.0"  rpy="${0.5*M_PI} ${0.0*M_PI} ${0.5*M_PI}" />
      <geometry>
        <mesh filename="package://tutorial_gazebo/meshes/folktruck_fork.stl" scale="${ratio} ${ratio} ${ratio}"/>
      </geometry>
    </collision>
  </link>

  <joint name="fork_joint" type="fixed">
    <origin xyz="0.0 0.0 -0.5" rpy="0 0 0"/>
    <axis xyz="0 0 1" />
    <limit effort="5000.0" lower="0.0" upper="2.85" velocity="0.2"/>
    <dynamics damping="1.0"/>
    <parent link="vplate"/>
    <child link="fork"/>
  </joint>


<link name="ltooth">
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="5" />
      <inertia ixx="0.021" ixy="0.0" ixz="0.0"
        iyy="0.021" iyz="0.0" izz="0.021" />
    </inertial>
    <visual>
      <origin xyz="0 0 0.0"  rpy="${0.5*M_PI} ${0.0*M_PI} ${0.5*M_PI}" />
      <geometry>
        <mesh filename="package://tutorial_gazebo/meshes/folktruck_forktooth.stl" scale="${ratio} ${ratio} ${ratio}"/>
      </geometry>
      <material name="orange" />
    </visual>

    <collision>
      <origin xyz="0 0 0.0"  rpy="${0.5*M_PI} ${0.0*M_PI} ${0.5*M_PI}" />
      <geometry>
        <mesh filename="package://tutorial_gazebo/meshes/folktruck_forktooth.stl" scale="${ratio} ${ratio} ${ratio}"/>
      </geometry>
    </collision>
  </link>

  <joint name="ltooth_joint" type="fixed">
    <origin xyz="0.0 0.0 0.0" rpy="0 0 0"/>
    <axis xyz="0 -1 0" />
    <limit effort="500.0" lower="0.0" upper="0.5" velocity="0.2"/>
    <dynamics damping="1.0"/>
    <parent link="fork"/>
    <child link="ltooth"/>
  </joint>


<link name="rtooth">
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="5" />
      <inertia ixx="0.021" ixy="0.0" ixz="0.0"
        iyy="0.021" iyz="0.0" izz="0.021" />
    </inertial>
    <visual>
      <origin xyz="0 0 0.0"  rpy="${0.5*M_PI} ${0.0*M_PI} ${0.5*M_PI}" />
      <geometry>
        <mesh filename="package://tutorial_gazebo/meshes/folktruck_forktooth.stl" scale="${ratio} ${ratio} ${ratio}"/>
      </geometry>
      <material name="orange" />
    </visual>

    <collision>
      <origin xyz="0 0 0.0"  rpy="${0.5*M_PI} ${0.0*M_PI} ${0.5*M_PI}" />
      <geometry>
        <mesh filename="package://tutorial_gazebo/meshes/folktruck_forktooth.stl" scale="${ratio} ${ratio} ${ratio}"/>
      </geometry>
    </collision>
  </link>

  <joint name="rtooth_joint" type="fixed">
    <origin xyz="0.0 -1.45 0.0" rpy="0 0 0"/>
    <axis xyz="0 1 0" />
    <limit effort="500.0" lower="0.0" upper="0.5" velocity="0.2"/>
    <dynamics damping="1.0"/>
    <parent link="fork"/>
    <child link="rtooth"/>
    <!--mimic joint="ltooth_joint" multiplier="1" offset="0"/--> 
  </joint>

<link name="plate">
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="0.1" />
      <inertia ixx="0.001" ixy="0.0" ixz="0.0"
        iyy="0.001" iyz="0.0" izz="0.001" />
    </inertial>
    <visual>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <geometry>
        <box size="1.8 0.12 0.04" />
      </geometry>
      <material name="orange" />
    </visual>

    <collision>
      <origin xyz="0 0 0.5" rpy="0 0 0" />
      <geometry>
        <sphere radius="0.0035" />
      </geometry>
    </collision>
  </link>

  <joint name="plate_joint" type="fixed">
    <origin xyz="-0.45 0 -0.25" rpy="0 0 0"/>
    <parent link="vplate"/>
    <child link="plate"/>
  </joint>

  <link name="camera_pod_link">
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="0.1" />
      <inertia ixx="0.001" ixy="0.0" ixz="0.0"
        iyy="0.001" iyz="0.0" izz="0.001" />
    </inertial>
 <visual>
      <origin xyz="0 0 0.0" rpy="0 0 0"/>
      <geometry>
        <cylinder  radius="0.03" length = "3.75"/>
      </geometry>
      <material name="orange" />
    </visual>

    <collision>
      <origin xyz="-0.05 0 0.5" rpy="0 0 0" />
      <geometry>
        <cylinder  radius="0.0035" length = "0.2"/>
      </geometry>
    </collision>
   
  </link>

  <joint name="camera_pod_joint" type="fixed">
    <origin xyz="0.4 0 1.55" rpy="0 0 0"/>
    <parent link="vplate"/>
    <child link="camera_pod_link"/>
  </joint>


  <link name="camera_link">
    <visual>
     <origin xyz="0 0 0" rpy="0 0 ${M_PI/2}"/>
      <geometry>
       <mesh filename="package://tutorial_gazebo/meshes/kinect.dae" scale="1.5 1.5 1.5"/>
      </geometry>
    </visual>
	  <collision>
	    <origin xyz="-0.05 0 0" rpy="0 0 ${M_PI/2}"/>
	    <geometry>
	      <box size="0.001 0.001 0.001"/>
	    </geometry>
	  </collision>
    <inertial>
      <mass value="0.5" />
      <origin xyz="0 0 0" />
      <inertia ixx="0.1" ixy="0.0" ixz="0.0"
               iyy="0.1" iyz="0.0"
               izz="0.1" />
    </inertial>
  </link>

  <joint name="camera_joint" type="fixed">
    <origin xyz="0.0 0.0 -0.0" rpy="0 0 0"/>
    <axis xyz="0 0 1" />
    <limit effort="50.0" lower="-0.20" upper="1.60" velocity="0.2"/>
    <dynamics damping="1.0"/>
    <parent link="camera_pod_link"/>
    <child link="camera_link"/>
  </joint>

   <link name="laser_visu">
    <inertial>
      <origin xyz="0 0 0"/>
      <mass value="0.01" />
      <inertia ixx="0.001" ixy="0.0" ixz="0.0"
        iyy="0.001" iyz="0.0" izz="0.001" />
    </inertial>
    <visual>
      <origin xyz="0 0 0" rpy="0 0 0" />
      <geometry>
        <mesh filename="package://tutorial_gazebo/meshes/hokuyo_utm_30lx.dae" scale="0.757 0.757 0.757"/> 
      </geometry>
    </visual>
  </link>

  <joint name="laser_visu_joint" type="fixed">
    <origin xyz="-1.225 0 0.98" rpy="0 0 ${M_PI}" />
    <parent link="base_visu_link" />
    <child link="laser_visu" />
  </joint>

   <link name="laser">
    <visual>
      <geometry>
        <box size="0.004 0.004 0.002" />
      </geometry>
    </visual>
    <inertial>
      <mass value="0.001" />
      <origin xyz="0 0 0" />
      <inertia ixx="0.0001" ixy="0.0" ixz="0.0"
        iyy="0.0001" iyz="0.0"
        izz="0.0001" />
    </inertial>
  </link>

  <joint name="laser_joint" type="fixed">
    <origin xyz="-1.225 0 0.98" rpy="0 0 ${M_PI}" />
    <parent link="base_visu_link" />
    <child link="laser" />
  </joint>


 <link name="laser2">
    <visual>
      <geometry>
        <mesh filename="package://tutorial_gazebo/meshes/hokuyo_utm_30lx.dae" scale="0.757 0.757 0.757"/> 
      </geometry>
    </visual>
    <inertial>
      <mass value="0.001" />
      <origin xyz="0 0 0" />
      <inertia ixx="0.0001" ixy="0.0" ixz="0.0"
        iyy="0.0001" iyz="0.0"
        izz="0.0001" />
    </inertial>
  </link>

  <joint name="laser2_joint" type="fixed">
    <origin xyz="-2.8 0 -0.20" rpy="0 0 1.5707" />
    <parent link="base_visu_link" />
    <child link="laser2" />
  </joint>

  <transmission name="tran0">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="vplate_joint"/>
    <actuator name="motor0">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran1">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="fork_joint"/>
    <actuator name="motor1">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran2">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="ltooth_joint"/>
    <actuator name="motor2">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran3">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="rtooth_joint"/>
    <actuator name="motor3">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

  <transmission name="tran4">
    <type>transmission_interface/SimpleTransmission</type>
    <joint name="camera_joint"/>
    <actuator name="motor4">
      <hardwareInterface>EffortJointInterface</hardwareInterface>
      <mechanicalReduction>1</mechanicalReduction>
    </actuator>
  </transmission>

<!--Painting color for all the joints-->
  <gazebo reference="f_wheel_link">
   <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="r_wheel_link">
   <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="base_footprint">
   <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="base_visu_link">
   <material>Gazebo/Orange</material>
  </gazebo>
  <gazebo reference="center_box">
   <material>Gazebo/Orange</material>
  </gazebo>
  <gazebo reference="plate">
   <material>Gazebo/Gold</material>
  </gazebo>
  <gazebo reference="vplate">
   <material>Gazebo/Grey</material>
  </gazebo>
  <gazebo reference="lcynldr">
   <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="rcynldr">
   <material>Gazebo/Black</material>
  </gazebo>  
  <gazebo reference="camera_pod_link">
   <material>Gazebo/FlatBlack</material>
  </gazebo>
  <gazebo reference="fork">
   <material>Gazebo/FlatBlack</material>
  </gazebo>
  <gazebo reference="ltooth">
    <kp>1000000.0</kp>
    <kd>1</kd>
    <mu1>10000000.0</mu1>
    <mu2>10000000.0</mu2>
    <fdir1>1 1 0</fdir1>
    <maxVel>1.0</maxVel>
    <maxContacts>100</maxContacts>
    <minDepth>0.00</minDepth>
   <material>Gazebo/Black</material>
  </gazebo>
  <gazebo reference="rtooth">
    <kp>1000000.0</kp>
    <kd>1.0</kd>
    <mu1>10000000.0</mu1>
    <mu2>10000000.0</mu2>
    <fdir1>1 1 0</fdir1>
    <maxVel>1.0</maxVel>
    <minDepth>0.00</minDepth>    
    <maxContacts>100</maxContacts>
   <material>Gazebo/Black</material>
  </gazebo>
  </xacro:macro>
</robot>
